File: backend/app/schemas/node.py
from pydantic import BaseModel, ConfigDict, Field


class RealityKeysResponse(BaseModel):
    private_key: str
    public_key: str
    short_id: str


class SniProfileBase(BaseModel):
    carrier: str = Field(..., min_length=1, max_length=100)
    domain: str = Field(..., min_length=1, max_length=255)
    is_active: bool = True


class SniProfileCreate(SniProfileBase):
    pass


class SniProfileUpdate(BaseModel):
    carrier: str | None = Field(default=None, min_length=1, max_length=100)
    domain: str | None = Field(default=None, min_length=1, max_length=255)
    is_active: bool | None = None


class SniProfileResponse(SniProfileBase):
    id: int
    node_id: int

    model_config = ConfigDict(from_attributes=True)


class NodeBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    host: str = Field(..., min_length=1, max_length=255)
    location: str = Field(default="", max_length=100)
    flag: str = Field(default="🌐", max_length=10)
    grpc_port: int = Field(default=10085, ge=1, le=65535)
    inbound_port: int = Field(default=443, ge=1, le=65535)


class NodeCreate(NodeBase):
    reality_private_key: str | None = None
    reality_public_key: str | None = None
    reality_short_id: str | None = None
    sni_profiles: list[SniProfileCreate] = []


class NodeUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    host: str | None = Field(default=None, min_length=1, max_length=255)
    location: str | None = Field(default=None, max_length=100)
    flag: str | None = Field(default=None, max_length=10)
    grpc_port: int | None = Field(default=None, ge=1, le=65535)
    inbound_port: int | None = Field(default=None, ge=1, le=65535)
    reality_private_key: str | None = None
    reality_public_key: str | None = None
    reality_short_id: str | None = None
    is_active: bool | None = None


class NodeResponse(NodeBase):
    id: int
    reality_private_key: str
    reality_public_key: str
    reality_short_id: str
    is_active: bool
    sni_profiles: list[SniProfileResponse] = []

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/node.ts
export interface SniProfile {
  id: number;
  node_id: number;
  carrier: string;
  domain: string;
  port: number;
  is_active: boolean;
}

export interface SniProfileCreate {
  carrier: string;
  domain: string;
  port?: number;
  is_active?: boolean;
}

export interface SniProfileUpdate {
  carrier?: string;
  domain?: string;
  port?: number;
  is_active?: boolean;
}

export interface NodeItem {
  id: number;
  name: string;
  host: string;
  location: string;
  flag: string;
  region_id?: number;
  grpc_port: number;
  inbound_port: number;
  reality_private_key: string;
  reality_public_key: string;
  reality_short_id: string;
  max_subscriptions: number;
  active_subscriptions_count?: number;
  is_active: boolean;
  sni_profiles: SniProfile[];
}

export interface NodeCreate {
  name: string;
  host: string;
  region_id?: number;
  location?: string;
  flag?: string;
  grpc_port?: number;
  inbound_port?: number;
  max_subscriptions?: number;
  reality_private_key?: string;
  reality_public_key?: string;
  reality_short_id?: string;
  sni_profiles?: SniProfileCreate[];
}

export interface NodeUpdate {
  name?: string;
  host?: string;
  region_id?: number;
  location?: string;
  flag?: string;
  grpc_port?: number;
  inbound_port?: number;
  max_subscriptions?: number;
  reality_private_key?: string;
  reality_public_key?: string;
  reality_short_id?: string;
  is_active?: boolean;
}

export interface RealityKeys {
  private_key: string;
  public_key: string;
  short_id: string;
}

export interface RegionStatus {
  flag: string;
  location: string;
  total_nodes: number;
  active_nodes: number;
  total_capacity: number;
  active_subscriptions: number;
  available_slots: number;
  is_sold_out: boolean;
}

